### PR TITLE
Add pluralize and singularize

### PR DESCRIPTION
**Pull request by tippenein**
Defines a similar interface to [activesupport inflections](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/inflections.rb)

This PR also makes some assumptions about what you want to include in this library.

I've used `pcre-light` regex library and copied over all the default mappings from activesupport. Perhaps you'd want to define these matchers with Megaparsec and skip a dependency?

Some type aliases have been used (`Singular`, `Plural`, `RegexReplace`, etc) which may or may not actually help (besides documentation (?))

I'd be cool with just moving this to a separate library if you don't use it here.

### CODE REVIEW

**Comment by jsl** [comment]
Hi @tippenein thanks very much for this contribution!

I think that I would rather use something like Megaparsec, as you mentioned. In fact, @mrkkrp helped with some of the cleanup in this library, so I'm guessing he would like to see the same thing. @mrkkrp do you have any thoughts on this PR?

**Comment by tippenein** [comment]
The more I think about it, the more gross regex seems. If you have an idea for how the user would interface with these parsers I can port over the relevant matching code.

File: Text/Inflections/Countable.hs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  matched : grouping <- regexMatch t reg
+  return $ (t `without` matched) <> groupReplace grouping rep
+  where
+    without t1 t2 = if t2 == "" then t1 else T.replace t2 "" t1
+
+groupReplace :: [T.Text] -> T.Text -> T.Text
+groupReplace g rep =
+  case g of
+    [] -> rep
+    (g':_) -> g' <> additions
```
Comment (mrkkrp) [inline Text/Inflections/Countable.hs]: This is going to be quite inefficient. Can we use replace `<>` with `:` here?

File: Text/Inflections/Countable.hs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+regexMatch :: T.Text -> Regex -> Maybe [T.Text]
+regexMatch t r = do
+  bs <- match r (encodeUtf8 t) []
+  return $ fmap decodeUtf8 bs
+
+regexPattern :: T.Text -> Maybe Regex
+regexPattern pat = toMaybe $ compileM (encodeUtf8 pat) [caseless]
+  where toMaybe = either (const Nothing) Just
+
+headMaybe :: [a] -> Maybe a
```
Comment (mrkkrp) [inline Text/Inflections/Countable.hs]: This is called `listToMaybe`, it lives in `Data.Maybe`.

File: Text/Inflections/Countable.hs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  where
+    mapping = defaultIrregulars ++ defaultUncountables ++ defaultSingulars
+
+-- | pluralize a word given a custom mapping.
+-- Build the [Inflection] with a combination of
+-- `makeUncountableMapping` `makeIrregularMapping` `makeMatchMapping`
+pluralizeWith :: [Inflection] -> T.Text -> T.Text
+pluralizeWith mapping t = fromMaybe t $ headMaybe matches
+  where
+    matches = catMaybes $ fmap (pluralLookup t) (Prelude.reverse mapping)
```
Comment (tippenein) [inline Text/Inflections/Countable.hs]: I believe it was a carry over from when I wasn't importing `Data.Text` qualified; will change.

**Comment by mrkkrp** [comment]
OK, there are two more things we need to be able to merge this:

1. We need to replace `pcre-light` with something. The reason for this is that `pcre-light` is a biding to C library `libpcre`, not pure Haskell library. This makes impossible to use `inflections` without an external C library which is a high price with respect to compatibility and portability. Could you please try [this Haskell implementation](https://hackage.haskell.org/package/regex-applicative) of regular expressions by Roman Cheplyaka. The drawback here is that it works with `String`s, which may be slower. Tell me what you think.

2. We need to fix the build. You could also squash all the commits together, but I can do that myself as well.

**Comment by tippenein** [comment]
Long overdue update:
The problem I see is that you'll need to create some language to convert the regex text `("(octop)us", "\\1i")` to an actual regex-applicative type, then run it on input text and convert back to text. I didn't see a way around having to map regex grammar to applicative grammar (which is the time-consuming part if I understand correctly)

Unfortunately, this isn't something I have time for at the moment. I've been using my `pcre-light` [version](https://github.com/tippenein/countable-inflections) so my need for this has decreased significantly. FWIW I still think this is worth doing and actually would be fun to do the regex -> applicative dsl-thingy. Perhaps I'll pick it up later, but you deserved an update on what was happening ;)

Cheers

**Comment by elcuy** [comment]
I will close this for now since it's been stale for so long. Will retake it if we need it in the future

**Pull request closed without merging**